File: code/ch16/16.5.1.sudoku-solver.py
class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        col = [[False] * 9 for i in range(9)]
        row = [[False] * 9 for i in range(9)]
        place = [[False] * 9 for i in range(9)]
        for i in range(9):
            for j in range(9):
                if board[i][j] != ".":
                    num = int(board[i][j]) - 1
                    col[j][num] = True
                    row[i][num] = True
                    place[i // 3 * 3 + j // 3][num] = True

        def dfs(i: int, j: int) -> boolean:
            if j == 9:
                return dfs(i + 1, 0)
            if i == 9:
                return True
            if board[i][j] != ".":
                return dfs(i, j + 1)
            for k in range(9):
                if (
                    col[j][k] == True
                    or row[i][k] == True
                    or place[i // 3 * 3 + j // 3][k] == True
                ):
                    continue
                board[i][j] = str(k + 1)
                col[j][k] = True
                row[i][k] = True
                place[i // 3 * 3 + j // 3][k] = True
                if dfs(i, j + 1) == True:
                    return True
                board[i][j] = "."
                col[j][k] = False
                row[i][k] = False
                place[i // 3 * 3 + j // 3][k] = False
            return False

        dfs(0, 0)

File: code/ch16/16.5.1.sudoku-solver.java
class Solution {
    
    boolean[][] col = new boolean[9][9];
    boolean[][] row = new boolean[9][9];
    boolean[][] place = new boolean[9][9];
    
    public void solveSudoku(char[][] board) {
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == '.') {
                    continue;
                }
                int num = board[i][j] - '1';
                row[i][num] = true;
                col[j][num] = true;
                place[i / 3 * 3 + j / 3][num] = true;
            }
        }
        dfs(board, 0, 0);
    }
    
    public boolean dfs(char[][] board, int i, int j) {
        if (j == 9) {
            return dfs(board, i + 1, 0);
        } else if (i == 9) {
            return true;
        } else if (board[i][j] != '.') {
            return dfs(board, i, j + 1);
        }

        for (int k = 0; k < 9; k++) {
            if (col[j][k] || row[i][k] || place[i / 3 * 3 + j / 3][k]) {
                continue;
            }
            board[i][j] = (char)('1' + k);
            col[j][k] = true;
            row[i][k] = true;
            place[i / 3 * 3 + j / 3][k] = true;
            if (dfs(board, i, j + 1)) {
                return true;
            }
            board[i][j] = '.';
            col[j][k] = false;
            row[i][k] = false;
            place[i / 3 * 3 + j / 3][k] = false;
        }
        return false;
    }
  
}